RIPS/rip-9995.md

```diff
@@ -0,0 +1,199 @@
+---
+rip: 7560
+title: Native Account Abstraction with EOF code sections
+description: A variant of RIP-7560 transactions relying on EOF Smart Contract Accounts
+author:
+discussions-to: https://ethereum-magicians.org/t/rip-7560-native-account-abstraction/16664
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-09-01
+requires: 7560, 3540
+---
+
+## Abstract
+
+This proposal describes a slight variation of the Native Account Abstraction design fully described in RIP-7560.
+This version's difference compared to the original proposal is in relying on features of "EVM Object Format" to
+distinguish between validation and execution code sections.
+
+## Motivation
+
+Talking about Full Native Account Abstraction, the fundamental idea any solution has to address is a
+mechanism for a Smart Contract Account to separate its validation and execution code sections.
+
+The RIP-7560 is build on the current Ethereum contract structure, and therefore has little choice but to rely
+on using some higher-level abstraction.
+
+In its current form, RIP-7560 transactions use Solidity method selectors in order to achieve this separation.
+
+This, however, is far from ideal as this approach "leaks" the concept from a programming language widely used
+in the EVM into the core design of the Ethereum protocol.
+
+While there is no purely technical reason to disallow it and there are already instances of Solidity code
+being "enshrined" in the Ethereum protocol, e.g. the validator deposit contract, such violation of abstraction levels
+often lead to unnecessary technical debt and are worth avoiding if possible.
+
+Additionally, method selectors provide very weak indication of the contract developer's decision to become a
+participant in Native Account Abstraction transaction flow.
+The chance of accidentally exposing a function with a colliding 4 byte
+method identifier that returns a valid 4 byte magic to indicate approval is pretty low, but a malicious developer
+can easily hide such a function giving it an innocent name, making it hard to spot a Native Account Abstraction entity.
+
+This issue to some extent is also present in ERC-4337.
+
+As an alternative, if Native Account Abstraction is to be implemented in coordination with EIP-3540,
+relying on the concept of "code sections" it introduces is a superior approach.
+
+## Specification
+
+### Validation and PostTransaction code sections
+
+The EOF format describes a separation of 'code' and 'data' sections,
+marked with `kind_code (0x02)` and `kind_data (0x04)` kind markers.
+
+The contract that may have a role in an Account Abstraction transaction, either as a Sender, a Paymaster or a Deployer,
+has to contain a section marked with the following kind marker:
+
+```
+kind_abstraction = 0x05
```

Comment:
`kind_aggregator_validation` also makes sense.
As for factory, it is less important, since we don't generate a calldata for it: we pass everything the wallet generates for it.
(and maybe more importantly: factory is doesn't validate anything: it runs in the context of validation, simply because we must create before being able to call validateUserOp